### PR TITLE
[Mapping] minor internal cleanup

### PR DESCRIPTION
**Pull request by philbucher**
Hi @jcotela 
In this PR I clean some not-so-nice things that I didn't know how to do better at the time.

Also I am finally putting some debugging checks into the debug compilation (I never had any of these errors in production, but they are very handy for developing so I don't wanna remove them)

### CODE REVIEW

File: applications/MappingApplication/custom_utilities/interface_object_manager_base.h [hunk 1]
```diff
@@ -300,7 +297,7 @@ class InterfaceObjectManagerBase
     // **********************************************************************
     // ** InterfaceObjectManagerSerial and InterfaceObjectManagerParallel **
     template <typename T>
-    void FillBufferWithValues(std::function<T(InterfaceObject*, const std::vector<double>&)> FunctionPointer,
+    void FillBufferWithValues(std::function<T(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointer,
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_base.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/interface_object_manager_base.h [hunk 2]
```diff
@@ -315,13 +312,13 @@ class InterfaceObjectManagerBase
...
         {
-            rBuffer[i] = FunctionPointer(boost::get_pointer(interface_obj), mShapeFunctionValues.at(mCommRank)[i]);
+            rBuffer[i] = FunctionPointer(interface_obj, mShapeFunctionValues.at(mCommRank)[i]);
             ++i;
         }
     }
 
     template <typename T>
-    void ProcessValues(std::function<void(InterfaceObject*, T)> FunctionPointer,
+    void ProcessValues(std::function<void(InterfaceObject::Pointer, T)> FunctionPointer,
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_base.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/interface_object_manager_base.h [hunk 3]
```diff
@@ -355,25 +349,25 @@ class InterfaceObjectManagerBase
     }
 
     virtual void FillBufferWithValues(double* pBuffer, int& rBufferSize, const int CommPartner,
-                                      std::function<double(InterfaceObject*, const std::vector<double>&)> FunctionPointer)
+                                      std::function<double(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointer)
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_base.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/interface_object_manager_base.h [hunk 4]
```diff
@@ -355,25 +349,25 @@ class InterfaceObjectManagerBase
...
     {
         KRATOS_ERROR << "Base class function called!" << std::endl;
     }
 
     virtual void FillBufferWithValues(double* pBuffer, int& rBufferSize, const int CommPartner,
-                                      std::function<array_1d<double, 3>(InterfaceObject*, const std::vector<double>&)> FunctionPointer)
+                                      std::function<array_1d<double, 3>(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointer)
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_base.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/interface_object_manager_base.h [hunk 5]
```diff
@@ -355,25 +349,25 @@ class InterfaceObjectManagerBase
...
     {
         KRATOS_ERROR << "Base class function called!" << std::endl;
     }
 
     virtual void ProcessValues(const double* pBuffer, const int BufferSize, const int CommPartner,
-                               std::function<void(InterfaceObject*, double)> FunctionPointer)
+                               std::function<void(InterfaceObject::Pointer, double)> FunctionPointer)
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_base.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/interface_object_manager_base.h [hunk 6]
```diff
@@ -355,25 +349,25 @@ class InterfaceObjectManagerBase
...
     {
         KRATOS_ERROR << "Base class function called!" << std::endl;
     }
 
     virtual void ProcessValues(const double* pBuffer, const int BufferSize, const int CommPartner,
-                               std::function<void(InterfaceObject*, array_1d<double, 3>)> FunctionPointer)
+                               std::function<void(InterfaceObject::Pointer, array_1d<double, 3>)> FunctionPointer)
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_base.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/interface_object_manager_parallel.h [hunk 7]
```diff
@@ -375,7 +372,7 @@ class InterfaceObjectManagerParallel : public InterfaceObjectManagerBase
     }
 
     void FillBufferWithValues(double* pBuffer, int& rBufferSize, const int CommPartner,
-                              std::function<double(InterfaceObject*, const std::vector<double>&)> FunctionPointer) override
+                              std::function<double(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointer) override
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_parallel.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/interface_object_manager_parallel.h [hunk 8]
```diff
@@ -386,21 +383,17 @@ class InterfaceObjectManagerParallel : public InterfaceObjectManagerBase
...
-        if (rBufferSize != i)
-        {
-            KRATOS_ERROR << "size mismatch" << std::endl;
-        }
+        KRATOS_DEBUG_ERROR_IF_NOT(rBufferSize == i) << "size mismatch" << std::endl;
     }
 
     void FillBufferWithValues(double* pBuffer, int& rBufferSize, const int CommPartner,
-                              std::function<array_1d<double, 3>(InterfaceObject*, const std::vector<double>&)> FunctionPointer) override
+                              std::function<array_1d<double, 3>(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointer) override
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_parallel.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/interface_object_manager_parallel.h [hunk 9]
```diff
@@ -424,46 +417,36 @@ class InterfaceObjectManagerParallel : public InterfaceObjectManagerBase
...
-        if (rBufferSize != i * 3)
-        {
-            KRATOS_ERROR << "size mismatch" << std::endl;
-        }
+        KRATOS_DEBUG_ERROR_IF_NOT(rBufferSize == i * 3) << "size mismatch" << std::endl;
     }
 
     void ProcessValues(const double* pBuffer, const int BufferSize, const int CommPartner,
-                       std::function<void(InterfaceObject*, double)> FunctionPointer) override
+                       std::function<void(InterfaceObject::Pointer, double)> FunctionPointer) override
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_parallel.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/interface_object_manager_parallel.h [hunk 10]
```diff
@@ -424,46 +417,36 @@ class InterfaceObjectManagerParallel : public InterfaceObjectManagerBase
...
         for (int i = 0; i < BufferSize; ++i)
         {
-            FunctionPointer(boost::get_pointer(interface_objects[i]), pBuffer[i]);
+            FunctionPointer(interface_objects[i], pBuffer[i]);
         }
     }
 
     void ProcessValues(const double* pBuffer, const int BufferSize, const int CommPartner,
-                       std::function<void(InterfaceObject*, array_1d<double, 3>)> FunctionPointer) override
+                       std::function<void(InterfaceObject::Pointer, array_1d<double, 3>)> FunctionPointer) override
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object_manager_parallel.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/mapper_communicator.h [hunk 11]
```diff
@@ -201,22 +201,22 @@ class MapperCommunicator
         InvokeSearch(mInitialSearchRadius, mMaxSearchIterations);
     }
 
-    virtual void TransferVariableData(std::function<double(InterfaceObject*, const std::vector<double>&)> FunctionPointerOrigin,
-                                      std::function<void(InterfaceObject*, double)> FunctionPointerDestination,
+    virtual void TransferVariableData(std::function<double(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointerOrigin,
+                                      std::function<void(InterfaceObject::Pointer, double)> FunctionPointerDestination,
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/mapper_communicator.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/mapper_communicator.h [hunk 12]
```diff
@@ -201,22 +201,22 @@ class MapperCommunicator
...
                                       const Variable<double>& rOriginVariable)
     {
         ExchangeDataLocal(FunctionPointerOrigin, FunctionPointerDestination);
     }
 
-    virtual void TransferVariableData(std::function<array_1d<double, 3>(InterfaceObject*, const std::vector<double>&)> FunctionPointerOrigin,
-                                      std::function<void(InterfaceObject*, array_1d<double, 3>)> FunctionPointerDestination,
+    virtual void TransferVariableData(std::function<array_1d<double, 3>(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointerOrigin,
+                                      std::function<void(InterfaceObject::Pointer, array_1d<double, 3>)> FunctionPointerDestination,
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/mapper_communicator.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/mapper_communicator.h [hunk 13]
```diff
@@ -201,22 +201,22 @@ class MapperCommunicator
...
                                       const Variable< array_1d<double, 3> >& rOriginVariable)
     {
         ExchangeDataLocal(FunctionPointerOrigin, FunctionPointerDestination);
     }
 
-    virtual void TransferVariableData(std::function<void*(InterfaceObject*, const std::vector<double>&)> FunctionPointerOrigin,
-                                      std::function<void(InterfaceObject*, void*)> FunctionPointerDestination )
+    virtual void TransferVariableData(std::function<void*(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointerOrigin,
+                                      std::function<void(InterfaceObject::Pointer, void*)> FunctionPointerDestination )
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/mapper_communicator.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/mapper_communicator.h [hunk 14]
```diff
@@ -319,8 +319,8 @@ class MapperCommunicator
     ///@{
 
     template <typename T>
-    void ExchangeDataLocal(std::function<T(InterfaceObject*, const std::vector<double>&)> FunctionPointerOrigin,
-                           std::function<void(InterfaceObject*, T)> FunctionPointerDestination)
+    void ExchangeDataLocal(std::function<T(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointerOrigin,
+                           std::function<void(InterfaceObject::Pointer, T)> FunctionPointerDestination)
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/mapper_communicator.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/mapper_mpi_communicator.h [hunk 15]
```diff
@@ -130,8 +130,8 @@ class MapperMPICommunicator : public MapperCommunicator
         mIntegrationMethodDestination = IntegrationMethodDestination;
     }
 
-    void TransferVariableData(std::function<double(InterfaceObject*, const std::vector<double>&)> FunctionPointerOrigin,
-                              std::function<void(InterfaceObject*, double)> FunctionPointerDestination,
+    void TransferVariableData(std::function<double(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointerOrigin,
+                              std::function<void(InterfaceObject::Pointer, double)> FunctionPointerDestination,
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/mapper_mpi_communicator.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/mapper_mpi_communicator.h [hunk 16]
```diff
@@ -142,8 +142,8 @@ class MapperMPICommunicator : public MapperCommunicator
         MPI_Barrier(MPI_COMM_WORLD);
     }
 
-    void TransferVariableData(std::function<array_1d<double, 3>(InterfaceObject*, const std::vector<double>&)> FunctionPointerOrigin,
-                              std::function<void(InterfaceObject*, array_1d<double, 3>)> FunctionPointerDestination,
+    void TransferVariableData(std::function<array_1d<double, 3>(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointerOrigin,
+                              std::function<void(InterfaceObject::Pointer, array_1d<double, 3>)> FunctionPointerDestination,
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/mapper_mpi_communicator.h]: Pass std::function<...> by reference

File: applications/MappingApplication/custom_utilities/mapper_mpi_communicator.h [hunk 17]
```diff
@@ -293,8 +293,8 @@ class MapperMPICommunicator : public MapperCommunicator
     }
 
     template <typename T>
-    void ExchangeDataRemote(std::function<T(InterfaceObject*, const std::vector<double>&)> FunctionPointerOrigin,
-                            std::function<void(InterfaceObject*, T)> FunctionPointerDestination)
+    void ExchangeDataRemote(std::function<T(InterfaceObject::Pointer, const std::vector<double>&)> FunctionPointerOrigin,
+                            std::function<void(InterfaceObject::Pointer, T)> FunctionPointerDestination)
```
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/mapper_mpi_communicator.h]: Pass std::function<...> by reference

**Review by jcotela — CHANGES_REQUESTED** [review]
Good cleanup overall, but I'd follow the recommendation of Codacy and pass the std::function<...> objects by reference. I had to look it up, but it seems that std::function objects may be relatively heavy (they are typically not so big, but they can be if the wrapped function requires it).

**Comment by philbucher** [comment]
I was already on that, thx @jcotela 
Now the function pointers are being passed by const reference

**Review by jcotela — APPROVED** [review]
Thanks!

**Pull request merged**